### ISSUE TITLE
Major water updates

### ISSUE THREAD

**Issue by redwyre** [StrangeLoopGames/EcoIssues#19390]
Started working on this but it's too much for 9.2 so writing out my plans here.

**Splitting WaterBlock**
`WaterBlock` currently is used for every kind of water, and uses extra data to store the water amount per block. This means the block is instanced for every usage which uses a lot of memory, and requires many chunks to use extra data (enabled if any block uses it) which is also sent to the players.

To alleviate this issue, I propose splitting `WaterBlock` into three new sub-types:
* `SurfaceWaterBlock` - This works like `WaterBlock` does today, it stores the water level and uses extra data, and does flow calculations. This is used for the top layer of water.
* `UnderWaterBlock` - This block is used for water blocks that are completely full and under the surface layer, since they are always full there is no need to store the water level so this block can instanced once and used for all under water blocks. If it doesn't have a `SurfaceWaterBlock` directly abocve it, it replaces itself with a `SurfaceWaterBlock`.
* `FallingWaterBlock` - this block is mostly to fix visual issues with water falling from a water fall or a pipe. It's water is continually transferred directly down. This fixes the "boiling water" effect which is an effort to reproduce this within the limitations of the block builder.

**Water flow direction**
@MazonDel has been looking into water flow, so things like water wheels work with correct direction of water. Also buoyant objects (like tree trunks) and rendering can use directional information.

* Surface water can have a directionallity added, and use 4 out of the 8 bits of extra data. This leaves 16 levels of water which is more than enough for client side rendering.

**Comment by johnkslg** [comment]
This sounds great, and should leverage existing stuff well.
One thing you may consider is allowing something akin to flow speed, for faster moving rivers that are steeper, etc. Though its less important.

**Comment by redwyre** [comment]
@johnkslg yeah, with 4bits/16 values that lets up transfer up to 3 speeds, depends if we want any other info

**Comment by redwyre** [comment]
UnderWater optimisation has been split to https://github.com/StrangeLoopGames/EcoIssues/issues/19454 for 9.3
